Restructuring of Server Tests

BlobbyBob opened the pull request:
Fixes #473 and as a side effect probably also #421 and #461 

:sparkles: All tests are capsuled in a transaction
:wrench: Database is filled through a script and not a sql dump
:wrench: Improved Type Hinting
:wrench: Adjusted Docker setup to work with tests

BlobbyBob commented:
@dorthrithil Can you take a look, whether you like the new test design?
If so, I would port the complete testing data and all test cases to the new structure.

dorthrithil commented:
Look good, yes! I was also always annoyed with the fact, that changing the test data involved creating a new dump etc. This way it would be much easier.
Probably we would use some speed though, I think pgrestore is faster then creating every row individually? Not sure though, this is just an assumption. Also, I would sacrifice some speed for the pro sides of this approach. Now that we don't login before each test, we probably gain 50% speed anyway which presumably dwarfes the pgrestore timeloss

BlobbyBob commented:
> Probably we would use some speed though, I think pgrestore is faster then creating every row individually?

I don't know if my test scenario is somehow biased, but the small setup in this MR (`test_region_resources`) terminates instantly (< 200ms I suppose), while before it took like 1 second per test case. The issue was that pg_restore has been called for *every single test*, while now the database is now populated *once for all tests*.

Maybe I'll do some benchmarking afterwards, but I think the overall runtime will dramatically improve.

dorthrithil commented:
I think populating it only once will break some tests as many of them depend on the initial database state to work (e.g. checking the length of a list won't work if an element was deleted in the previous test). pytest is not guaranteed to always execute all tests in the same order, so there's also no straight forward way to work around this.

However, there could be a way around this using `pytest-ordering`? Do you have some additional ideas for this?

BlobbyBob commented:
Of course this would be a problem, but I have already thought of that:

> ✨ All tests are capsuled in a transaction

The idea is that every test gets its own transaction, which is rolled back in the end instead of being committed. I've used Django's `TransactionTestCase` as inspiration. Unfortunately, you need to write the logic yourself in Flask, but I think the current proposal for the [db_session fixture](https://github.com/LocalCrag/LocalCragApp/pull/474/commits/a178991e70318907a6713437c18fde55d5586fdb#diff-dd9c38840c14d9cc476d66287ba75aa454850ee048c04d4ea3f48e6748ad509eR51) should hopefully do it.

dorthrithil commented:
Ahh nice, didn't see the rollback when briefly looking over it. This way it will be really fast I suppose, nice.

BlobbyBob commented:
I am currently too lazy to setup the old testing setup, but here is the runtime statistic for the new one:  
`170 passed, 20 warnings in 5.10s`

dorthrithil commented on server/src/models/user.py:
```diff
@@ -28,7 +30,7 @@ class User(HasSlug, IsSearchable, BaseEntity):
     reset_password_hash_created = db.Column(db.DateTime(timezone=True), default=None, nullable=True)
     language = db.Column(db.String, nullable=False, server_default='de')
-    avatar_id = db.Column(UUID(), db.ForeignKey('files.id'), nullable=True)
+    avatar_id = db.Column(UUID(), db.ForeignKey('files.id', name="fk_user_avatar_id"), nullable=True)
```
Why is an explicit fk name necessary here?

dorthrithil commented on server/tests/conftest.py:
```diff
@@ -98,3 +204,416 @@ def clean_test_uploads():
+    file.thumbnail_xs = True
+    file.thumbnail_s = True
+    file.thumbnail_m = True
+    file.filename = "	f4947937541045fdade9cfdf5cdb5afa.png"
```
spaces `"	f`

BlobbyBob replied:
It was even a tab :D Probably copied it from the adminer web UI

dorthrithil commented on server/tests/test_crag_resources.py:
```diff
@@ -24,85 +24,79 @@ def test_successful_create_crag(client):
+    assert res['name'] == crag.name
+    assert res['slug'] == crag.slug
+    assert res['orderIndex'] == crag.order_index
+    # assert res['ascentCount'] == crag.ascent_count
```
unintended comment?

BlobbyBob replied:
For some reason derived access like `ascent_count` sometimes fail with an error that it is not executed in context of a request, while the `db_session` fixture should imply ann `app.context`. I did not find the reason behind, so I removed the few occasions of this issue. It must be noted, that all derived functionality is also tested on its own somewhere else, so I don't think that any bug will slip because of this.

dorthrithil commented on server/tests/test_line_resources.py:
```diff
@@ -654,39 +643,39 @@ def test_create_line_invalid_video_payload(client):
 
 
 def test_successful_get_lines(client):
     rv = client.get('/api/lines')
```
For other resources you used a new srtategy I like, e.g. `for r, a in zip(res, areas):` would be cool to have that in the lines tests as well

BlobbyBob replied:
done, for the others too

dorthrithil commented on server/tests/test_menu_page_resources.py:
```diff
@@ -1,44 +1,40 @@
-    assert res[1]['title'] == "Impressum"
-    assert res[1]['text'] == "<p>Hier steht ein Impressums Text.</p>"
-    assert res[0]['id'] == "f2b0606f-46f6-4012-be67-5315bba154d2"
+    assert isinstance(res[0]['id'], str)
```
Same as in the lines tests, load objects for assertions first

dorthrithil commented on server/tests/test_post_resources.py:
```diff
@@ -22,13 +16,13 @@ def test_successful_create_post(client):
+    res = rv.json
     assert len(res) == 2
-    assert res[0]['id'] == "e8b86cba-fdd6-43a7-aea3-3d3aabedf795"
+    assert isinstance(res[0]['id'], str)
```
same as lines, see above

dorthrithil commented on server/tests/test_sector_resources.py:
```diff
@@ -43,63 +43,61 @@ def test_successful_create_sector(client):
-    assert res[1]['ascentCount'] == 0
-    assert res[1]['secret'] == False
-    assert res[0]['id'] == "008478de-5e0b-41b3-abe7-571f758c189b"
+    assert isinstance(res[0]['id'], str)
```
same as in lines, see above

dorthrithil reviewed (changes requested):
Really good work, thank you! This will make the tests way easier to use and faster. I found only some minor points

dorthrithil commented:
I think you can also delete the test dumps

dorthrithil commented:
> I am currently too lazy to setup the old testing setup, but here is the runtime statistic for the new one on my PC: `170 passed, 20 warnings in 5.10s`

Nice! They need 27s on my machine currently (in the old version). Probably even faster then 5.1s now, I have to check this. Also curious to see the speed improvement in the pipeline

BlobbyBob commented:
All things should be fixed. I also added the `tests` directory to all of the linting tools.
Some of the test CI jobs however fail now, because I deleted the database dump as you suggested :sweat_smile: 
Do you want to adjust the CI configuration or should we restore the dump?

dorthrithil commented:
Pipelines are now running again with the new test setup
